python/refactor/nn/modules/encoder.py

```diff
@@ -0,0 +1,138 @@
+from .module import Module
+from .ke_embedding import KGEmbedding
+import torch as th
+import itertools
+
+# this is an interface so it will not have many parameters to be defined.
+class Encoder(Module):
+    def __init__(self, encoder_name):
+        super(Encoder, self).__init__(encoder_name)
+
+    def forward(self, data, gpu_id):
+        raise NotImplementedError
+
+class KGEEncoder(Encoder):
+    def __init__(self,
+                 hidden_dim,
+                 lr,
+                 optimizer,
+                 n_entity,
+                 n_relation,
+                 init_func,
+                 rel_parts=None, # ! rel_parts and cross_rels must be put here otherwise harm the general structure
+                 cross_rels=None,
+                 soft_rel_part=False,
+                 strict_rel_part=False,
+                 encoder_name='KGEEncoder',):
+        super(KGEEncoder, self).__init__(encoder_name)
+        self.entity_emb = KGEmbedding()
+        self.relation_emb = KGEmbedding()
+        self.entity_emb.init(lr=lr,
+                             num=n_entity,
+                             dim=hidden_dim,
+                             init_func=init_func[0],
+                             optimizer=optimizer)
+        self.relation_emb.init(lr=lr,
+                               num=n_relation,
+                               dim=hidden_dim,
+                               init_func=init_func[1],
+                               optimizer=optimizer)
+        self.rel_parts = rel_parts
+        self.cross_rels = cross_rels
+        self.soft_rel_part = soft_rel_part
+        self.strict_rel_part = strict_rel_part
+        self.init = False
+
+    def save(self, save_path):
+        self.entity_emb.save(save_path, 'entity_emb.npy')
+        self.relation_emb.save(save_path, 'relation_emb.npy')
+
+    def load(self, load_path):
+        self.entity_emb.load(load_path, 'entity_emb.npy')
+        self.relation_emb.load(load_path, 'relation_emb.npy')
```

Comment:
as we discussed last time, we should move relation_emb to KGE decoder, right?

Reply:
I believed we said that node_emb and relation_emb should be in encoder, and other stuff in decoder. I will double check with Xiang.